File: examples/Epitaxy/Epitaxy.py
import viennals as vls
import time

# Dimension
D = 3
vls.setNumThreads(8)

class EpitaxyVelocity(vls.VelocityField):
    def __init__(self, velocities):
        super().__init__()
        self.velocities = velocities
        self.R111 = 0.5
        self.R100 = 1.0
        # D is 3 in this context
        self.low = 0.5773502691896257 if D > 2 else 0.7071067811865476
        self.high = 1.0

    def getScalarVelocity(self, coord, material, normal, point_id):
        vel = max(abs(normal[0]), abs(normal[2]))
        factor = (self.R100 - self.R111) / (self.high - self.low)
        vel = (vel - self.low) * factor + self.R111

        if abs(normal[0]) < abs(normal[2]):
            vel *= 2.0

        mat_vel = 0.0
        if material < len(self.velocities):
            mat_vel = self.velocities[material]

        return vel * mat_vel

    def getVectorVelocity(self, coord, material, normal, point_id):
        return (0., 0., 0.)

def write_surface(domain, filename):
    mesh = vls.Mesh()
    vls.ToSurfaceMesh(domain, mesh).apply()
    vls.VTKWriter(mesh, filename).apply()

def main():
    # Set dimension to 3D
    vls.setDimension(D)

    # Parameters
    grid_delta = 0.03
    fin_width = 0.5
    fin_height = 0.2

    bounds = [-1.0, 1.0, -1.0, 1.0, -1.0, 1.0]
    boundary_conditions = [
        vls.BoundaryConditionEnum.REFLECTIVE_BOUNDARY,
        vls.BoundaryConditionEnum.REFLECTIVE_BOUNDARY,
        vls.BoundaryConditionEnum.INFINITE_BOUNDARY
    ]

    # Create Geometry
    mask = vls.Domain(bounds, boundary_conditions, grid_delta)
    plane = vls.Plane([0.0, 0.0, 0.0], [0.0, 0.0, 1.0])
    vls.MakeGeometry(mask, plane).apply()

    substrate = vls.Domain(bounds, boundary_conditions, grid_delta)
    min_point = [-fin_width / 2, -fin_width / 2, 0.0]
    max_point = [fin_width / 2, fin_width / 2, fin_height]
    box = vls.Box(min_point, max_point)
    vls.MakeGeometry(substrate, box).apply()

    vls.BooleanOperation(substrate, mask, vls.BooleanOperationEnum.UNION).apply()

    write_surface(mask, "mask.vtp")
    write_surface(substrate, "substrate.vtp")

    level_sets = [mask, substrate]
    vls.PrepareStencilLocalLaxFriedrichs(level_sets, [False, True])

    # Advection
    velocities = [0.0, -0.5]
    velocity_field = EpitaxyVelocity(velocities)

    advection = vls.Advect()
    for ls in level_sets:
        advection.insertNextLevelSet(ls)
    advection.setVelocityField(velocity_field)
    advection.setIntegrationScheme(vls.IntegrationSchemeEnum.STENCIL_LOCAL_LAX_FRIEDRICHS_1ST_ORDER)
    advection.setAdvectionTime(0.5)

    start_time = time.time()
    advection.apply()
    end_time = time.time()
    print(f"Epitaxy took {end_time - start_time}s")

    vls.FinalizeStencilLocalLaxFriedrichs(level_sets)

    write_surface(substrate, "epitaxy.vtp")

if __name__ == "__main__":
    main()

File: examples/Epitaxy/Epitaxy.cpp
#include <lsAdvect.hpp>
#include <lsBooleanOperation.hpp>
#include <lsDomain.hpp>
#include <lsMakeGeometry.hpp>
#include <lsToSurfaceMesh.hpp>
#include <lsVTKWriter.hpp>

#include <vcTimer.hpp>

using namespace viennals;
constexpr int D = 3;
using T = double;

// Implement Epitaxy velocity field
class epitaxy final : public VelocityField<T> {
  std::vector<double> velocities;
  static constexpr double R111 = 0.5;
  static constexpr double R100 = 1.;
  static constexpr double low =
      (D > 2) ? 0.5773502691896257 : 0.7071067811865476;
  static constexpr double high = 1.0;

public:
  epitaxy(std::vector<double> vel) : velocities(vel){};

  double getScalarVelocity(const std::array<T, 3> & /*coordinate*/,
                           int material, const std::array<T, 3> &normal,
                           unsigned long /* pointID */) final {
    double vel = std::max(std::abs(normal[0]), std::abs(normal[2]));
    constexpr double factor = (R100 - R111) / (high - low);
    vel = (vel - low) * factor + R111;
    if (std::abs(normal[0]) < std::abs(normal[2])) {
      vel *= 2.;
    }

    return vel *
           ((material < int(velocities.size())) ? velocities[material] : 0);
  }
};

void writeSurface(SmartPointer<Domain<T, D>> domain,
                  const std::string &filename) {
  auto mesh = Mesh<T>::New();
  ToSurfaceMesh<T, D>(domain, mesh).apply();
  VTKWriter<T>(mesh, filename).apply();
}

int main(int argc, char *argv[]) {

  // Create hole geometry
  double bounds[2 * D] = {-1.0, 1.0, -1.0, 1.0, -1.0, 1.0};
  BoundaryConditionEnum boundaryConditions[2 * D] = {
      BoundaryConditionEnum::REFLECTIVE_BOUNDARY,
      BoundaryConditionEnum::REFLECTIVE_BOUNDARY,
      BoundaryConditionEnum::INFINITE_BOUNDARY};
  const T gridDelta = 0.03;

  const T finWidth = 0.5;
  const T finHeight = 0.2;

  auto mask = Domain<T, D>::New(bounds, boundaryConditions, gridDelta);
  MakeGeometry<T, D>(mask, Plane<T, D>::New(VectorType<T, D>{0.0, 0.0, 0.0},
                                            VectorType<T, D>{0.0, 0.0, 1.0}))
      .apply();

  auto substrate = Domain<T, D>::New(bounds, boundaryConditions, gridDelta);
  T minPoint[D] = {-finWidth / 2, -finWidth / 2, 0.};
  T maxPoint[D] = {finWidth / 2, finWidth / 2, finHeight};
  MakeGeometry<T, D>(substrate, Box<T, D>::New(minPoint, maxPoint)).apply();
  BooleanOperation(substrate, mask, BooleanOperationEnum::UNION).apply();

  writeSurface(mask, "mask.vtp");
  writeSurface(substrate, "substrate.vtp");

  std::vector<SmartPointer<Domain<T, D>>> levelSets;
  levelSets.push_back(mask);
  levelSets.push_back(substrate);

  PrepareStencilLocalLaxFriedrichs<T, D>(levelSets,
                                         std::vector<bool>{false, true});

  auto velocityField = SmartPointer<epitaxy>::New(std::vector<double>{0., -.5});

  Advect<T, D> advectionKernel(levelSets, velocityField);
  advectionKernel.setSpatialScheme(
      SpatialSchemeEnum::STENCIL_LOCAL_LAX_FRIEDRICHS_1ST_ORDER);
  advectionKernel.setAdvectionTime(.5);

  Timer timer;
  timer.start();
  advectionKernel.apply();
  timer.finish();

  std::cout << "Epitaxy took " << timer.currentDuration / 1e9 << "s\n";

  FinalizeStencilLocalLaxFriedrichs<T, D>(levelSets);

  writeSurface(substrate, "epitaxy.vtp");
}